src/core/shaders/utils.js

```diff
@@ -1,32 +1,30 @@
-import ShaderCache from './ShaderCache';
+import ProgramCache from './ProgramCache';
 
 let programID = 1;
+const programCache = new ProgramCache();
 
-const shaderCache = new ShaderCache();
-
+/**
+ * Compile a webgl program.
+ * Use a cache to improve speed.
+ * 
+ * @param {WebGLRenderingContext} gl - The context where the program will be executed
+ * @param {string} glslVS - vertex shader code
+ * @param {string} glslFS - fragment shader code
+ */
 export function compileProgram(gl, glslVS, glslFS) {
+    const code = glslVS + glslFS;
+    if (programCache.has(gl, code)) {
+        return programCache.get(gl, code);
+    }
     const VS = compileShader(gl, glslVS, gl.VERTEX_SHADER);
     const FS = compileShader(gl, glslFS, gl.FRAGMENT_SHADER);
-    const program = gl.createProgram();
-    gl.attachShader(program, VS);
-    gl.attachShader(program, FS);
-    gl.linkProgram(program);
-    gl.deleteShader(VS);
-    gl.deleteShader(FS);
-    if (!gl.getProgramParameter(program, gl.LINK_STATUS)) {
-        throw new Error('Unable to link the shader program: ' + gl.getProgramInfoLog(program));
-    }
 
-    return {
-        program: program,
-        programID: programID++
-    };
+    const program = createProgram(gl, VS, FS);
+    programCache.set(gl, code, { program: program, programID: programID++ });
+    return { program, programID };
 }
 
 function compileShader(gl, sourceCode, type) {
-    if (shaderCache.has(gl, sourceCode)) {
```

Comment:
But, don't remove the shader (vertex/fragment) one!

This is still needed. A program may need to be recompiled, but not all parts may need to be recompiled. Actually, this is a very typical behavior since restyling modifies the fragment shader but the vertex shader is fixed.